tests/unit/s3/test_connection.py

```diff
@@ -396,6 +398,298 @@ def test_head_bucket_other(self):
+    b'<Region>us-east-2</Region><RequestId>asdf</RequestId>'
+    b'<HostId>asdf</HostId></Error>'
+) % RETRY_REGION_BYTES
+
+ILLEGAL_LOCATION_CONSTRAINT_SPECIFIED_REGION = (
+    b'<Error><Code>IllegalLocationConstraintException</Code>'
+    b'<Message>The %s location constraint is incompatible '
+    b'for the region specific endpoint this request was sent to.'
+    b'</Message><RequestId>asdf</RequestId>'
+    b'<HostId>asdf</HostId></Error>'
+) % RETRY_REGION_BYTES
+
+PERMANENT_REDIRECT = (
+    b'<Error><Code>PermanentRedirect</Code><Message>The bucket you '
+    b'are attempting to access must be addressed using the '
+    b'specified endpoint. Please send all future requests to this '
+    b'endpoint.</Message><Endpoint>'
+    b'bucket.s3.%s.amazonaws.com</Endpoint>'
+    b'<Bucket>bucket</Bucket><RequestId>asdf</RequestId>'
+    b'<HostId>asdf</HostId></Error>'
+) % RETRY_REGION_BYTES
+
+ILLEGAL_LOCATION_CONSTRAINT_UNSPECIFIED_REGION = (
+    b'<Error><Code>IllegalLocationConstraintException</Code>'
+    b'<Message>The unspecified location constraint is incompatible '
+    b'for the region specific endpoint this request was sent to.'
+    b'</Message><RequestId>asdf</RequestId>'
+    b'<HostId>asdf</HostId></Error>'
+)
+
+# This output cannot come from the API, but it's here to test
+# that strings not matching the regex handling this error are
+# handled gracefully.
+ILLEGAL_LOCATION_CONSTRAINT_MALFORMED = (
+    b'<Error><Code>IllegalLocationConstraintException</Code>'
+    b'<Message>Some random string.</Message><RequestId>asdf</RequestId>'
+    b'<HostId>asdf</HostId></Error>'
+)
+
+ERRORS_WITH_REGION_IN_BODY = (
+    (400, AUTHORIZATION_HEADER_MALFORMED),
+    (400, ILLEGAL_LOCATION_CONSTRAINT_SPECIFIED_REGION),
+    (301, PERMANENT_REDIRECT)
+)
+
+ERRORS_WITHOUT_REGION_IN_BODY = (
+    (400, ILLEGAL_LOCATION_CONSTRAINT_UNSPECIFIED_REGION),
+    (400, ILLEGAL_LOCATION_CONSTRAINT_MALFORMED)
+)
+
+
+class TestMakeRequestRetriesWithCorrectHost(AWSMockServiceTestCase):
+
+    def setUp(self):
+        self.connection = AWSAuthConnection('s3.amazonaws.com')
+
+        self.non_retriable_code = 404
+        self.retry_status_codes = [301, 400]
+        self.success_response = self.create_response(200)
+
+        self.default_host = 'bucket.s3.amazonaws.com'
+        self.retry_region = RETRY_REGION_BYTES.decode('utf-8')
+        self.default_retried_host = (
+            'bucket.s3.%s.amazonaws.com' % self.retry_region
+        )
+        self.test_headers = [('x-amz-bucket-region', self.retry_region)]
+
+    def test_non_retriable_status_returns_original_response(self):
+        error_response = self.create_response(self.non_retriable_code)
+        self.connection._mexe = mock.Mock(side_effect=[error_response])
+
+        response = self.connection.make_request(
+            'HEAD', '/', host=self.default_host)
+
+        self.assertEqual(response, error_response)
+        self.assertEqual(self.connection._mexe.call_count, 1)
```

Comment:
Summarizing what we discussed elsewhere for posterity: the HTTPRequest class doesn't implement __eq__, so assert_called_once_with doesn't check equality properly. Rather than hack my way around this, I just checked the host of the request instead, since that is what's relevant to the logic here.